hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/TransactionPoolSnapshotUtility.java

```diff
@@ -0,0 +1,275 @@
+import static com.hazelcast.jet.config.ProcessingGuarantee.EXACTLY_ONCE;
+import static com.hazelcast.jet.config.ProcessingGuarantee.NONE;
+import static com.hazelcast.jet.core.BroadcastKey.broadcastKey;
+import static java.util.Collections.singletonList;
+
+/**
+ * A utility to handle transactions where each local processor uses a pool of
+ * transactional resources alternately. This is needed if we can't reliably
+ * list transactions that were created by our processor and solves it by having
+ * a deterministic transaction IDs.
+ */
+public class TransactionPoolSnapshotUtility<TXN_ID extends TransactionId, RES extends TransactionalResource<TXN_ID>>
+        extends TwoPhaseSnapshotCommitUtility<TXN_ID, RES> {
+
+    private static final int TXN_PROBING_FACTOR = 5;
+
+    private final int poolSize;
+    private final List<TXN_ID> transactionIds;
+    private List<LoggingNonThrowingResource<TXN_ID, RES>> transactions;
+    private int activeTxnIndex;
+    private boolean activeTransactionUsed;
+    private TXN_ID preparedTxnId;
+    private LoggingNonThrowingResource<TXN_ID, RES> transactionToCommit;
+    private boolean flushed;
+    private boolean processorCompleted;
+    private boolean transactionsReleased;
+
+    /**
+     * @param createTxnIdFn input is {processorIndex, transactionIndex}
+     */
+    public TransactionPoolSnapshotUtility(
+            @Nonnull Outbox outbox,
+            @Nonnull Context procContext,
+            boolean isSource,
+            @Nonnull ProcessingGuarantee externalGuarantee,
+            int poolSize,
+            @Nonnull BiFunctionEx<Integer, Integer, TXN_ID> createTxnIdFn,
+            @Nonnull FunctionEx<TXN_ID, RES> createTxnFn,
+            @Nonnull ConsumerEx<TXN_ID> recoverAndCommitFn,
+            @Nonnull ConsumerEx<TXN_ID> recoverAndAbortFn
+    ) {
+        super(outbox, procContext, isSource, externalGuarantee, createTxnFn, recoverAndCommitFn,
+                processorIndex -> {
+                    for (int i = 0; i < adjustPoolSize(externalGuarantee, isSource, poolSize); i++) {
+                        TXN_ID txnId = createTxnIdFn.apply(processorIndex, i);
+                        LoggingUtil.logFine(procContext.logger(), "recover and abort %s", txnId);
+                        recoverAndAbortFn.accept(txnId);
+                    }
+                });
+
+        this.poolSize = adjustPoolSize(externalGuarantee, isSource, poolSize);
+        LoggingUtil.logFine(procContext.logger(), "Actual pool size used: %d", this.poolSize);
+        if (this.poolSize > 1) {
+            transactionIds = new ArrayList<>(this.poolSize);
+            for (int i = 0; i < this.poolSize; i++) {
+                transactionIds.add(createTxnIdFn.apply(procContext().globalProcessorIndex(), i));
+                assert i == 0 || !transactionIds.get(i).equals(transactionIds.get(i - 1)) : "two equal IDs generated";
+            }
+        } else {
+            transactionIds = singletonList(null);
```

Comment:
we create a list wit a `null` if poolsize=1 and then a single transaction is created with `null` transactionId. This means actually we don't use the transaction life-cycle right ?

Reply:
yes